### PR TITLE
fix(over window): fix over window predicate pushdown

### LINKED ISSUE #13653: bug(optimizer): `PredicatePushdown` across `OverWindow`
### Describe the bug

We may prematurely push a predicate down across an `OverWindow` node, producing an non-equivalent plan.

### Error message/log

_No response_

### To Reproduce

```
create table t (id int, cat varchar, rule varchar, at timestamptz);
insert into t values (1, 'foo', 'A', '2023-11-23T12:00:42Z'), (2, 'foo', 'B', '2023-11-23T12:01:15Z');
```

Correct (verified against PostgreSQL):
```
dev=> select rule, lag(rule) over (partition by cat order by at) from t;
 rule | lag 
------+-----
 A    | 
 B    | A
(2 rows)

dev=> select rule, lag(rule) over (partition by cat order by at) from t where rule = 'B';
 rule | lag 
------+-----
 B    | 
(1 row)

dev=> select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) as with_prev;
 rule | prev_rule 
------+-----------
 A    | 
 B    | A
(2 rows)

dev=> explain select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) as with_prev;
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 BatchExchange { order: [], dist: Single }
 └─BatchProject { exprs: [t.rule, first_value] }
   └─BatchOverWindow { window_functions: [first_value(t.rule) OVER(PARTITION BY t.cat ORDER BY t.at ASC ROWS BETWEEN 1 PRECEDING AND 1 PRECEDING)] }
     └─BatchExchange { order: [t.cat ASC, t.at ASC], dist: HashShard(t.cat) }
       └─BatchSort { order: [t.cat ASC, t.at ASC] }
         └─BatchScan { table: t, columns: [cat, rule, at] }
(6 rows)
```

Wrong:
```
dev=> select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) as with_prev where rule = 'B';
 rule | prev_rule 
------+-----------
 B    | 
(1 row)

dev=> explain select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) as with_prev where rule = 'B';
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 BatchExchange { order: [], dist: Single }
 └─BatchProject { exprs: [t.rule, first_value] }
   └─BatchOverWindow { window_functions: [first_value(t.rule) OVER(PARTITION BY t.cat ORDER BY t.at ASC ROWS BETWEEN 1 PRECEDING AND 1 PRECEDING)] }
     └─BatchExchange { order: [t.cat ASC, t.at ASC], dist: HashShard(t.cat) }
       └─BatchSort { order: [t.cat ASC, t.at ASC] }
         └─BatchFilter { predicate: (t.rule = 'B':Varchar) }
           └─BatchScan { table: t, columns: [cat, rule, at] }
(7 rows)
```

### Expected behavior

As in PostgreSQL:
```
test=# select rule, lag(rule) over (partition by cat order by at) from t where rule = 'B'; -- already correct
 rule | lag 
------+-----
 B    | 
(1 row)

test=# select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) as with_prev where rule = 'B';
 rule | prev_rule 
------+-----------
 B    | A
(1 row)
```

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Given the following queries:

```sql
select rule, lag(rule) over (partition by cat order by at) from t where rule = 'B'; -- Q1
select * from (select rule, lag(rule) over (partition by cat order by at) as prev_rule from t) where rule = 'B'; -- Q2
```

By their semantics, Q1 requires filtering rows read from `t`, while Q2 requires filtering rows after window function calculation. Previously, for Q2, our optimizer incorrectly pushed the predicate down before over window, which diverged from the expected semantics.

Thanks to @xiangjinwu's https://github.com/risingwavelabs/risingwave/issues/13653 for revealing such a long-existing bug. This PR fixes #13653.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_over_window.rs [hunk 1]
```diff
@@ -698,9 +698,21 @@ impl PredicatePushdown for LogicalOverWindow {
...
+                .iter()
+                .map(|arg| arg.index)
+                .chain(func.order_by.iter().map(|o| o.column_index))
+        });
+        let mut over_window_related_cols: FixedBitSet = window_func_input_refs
+            .chain(in_schema_len..out_schema_len)
+            .collect();
+        over_window_related_cols.grow(out_schema_len);
+
+        let (window_pred, other_pred) = predicate.split_disjoint(&over_window_related_cols);
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/logical_over_window.rs:715]: Considering this case, even if we use a `id` which should be not `over_window_related_cols` column, we should not push this filter through the over window.
```sql
 select * from (select id, rule, at, row_number() over (partition by cat order by at) as rank from t) x where id = 2;
```

Comment (stdrc) [reply src/frontend/src/optimizer/plan_node/logical_over_window.rs:715]: Then it seems that we cannot push any predicate through over window.

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/logical_over_window.rs:715]: However, partition only related filter could be pushed down

**Review by chenzl25 — APPROVED** [review]
LGTM, thanks!

**Review by st1page — APPROVED** [review]

**Review by xiangjinwu — APPROVED** [review]
Just need to merge main one more time ...

**Pull request merged**